### ISSUE TITLE
Bake in some helper classes

### ISSUE THREAD

**Issue by gregrickaby** [WebDevStudios/wd_s#326]
Starting a ticket to dump a list of potential helper classes brought up on the FED scrum by @efuller.

cc @jomurgel

**Comment by efuller** [comment]
So here is a bit of a brain dump from different discussions over on Teamsters. Again, just conversation starters!

### Overview
Without going too crazy, Teamsters has been trying to use presentational/helper classes to help with consistency and to maximize reuse. Taking this approach promotes consistency in regards to *global* things.

Current helpers that are already baked in like the `color` classes and `transition` variables are 👍 💯 . Let's keep those! I think we all agreed that removing the grid helpers was 👍 .

Along with baking in some additional helpers, perhaps we could introduce a `helpers` folder inside the `utilities` folder. This folder could house classes that might need to be created on a per-project basis and we could even think about moving some of the current helpers in there as well such as the `color` helper classes.

I personally really like the idea of creating helper classes on a per project basis for CSS solutions that *might* or *could* be used in different places across the project.

## Additional Helpers

### Spacing
Consistent spacing across a website is hard. In an effort to help with this, Teamsters have been working to control global spacing via pre-defined helper classes only. 

Jo's first implementation of creating these classes was to use a list of spacings (20, 30, 40, 50) and from that list create the helper classes using a Sass loop.

Below are some example classes that could be interchanged with padding or margins.
```
.pad-100-0 {};
.pad-50-0 {};
.pad-0-15 {};
.pad-top-50 {};
.pad-bottom-0 {}; 
```
We could work to agree on a core set of spacing classes to start with, then add to as needed.

I'd really like to have @jomurgel weigh in here as well as anyone else who has ideas on potential helpers that could help us be more efficient.

In the end, I think it would be great to have a core set of helpers baked in that we can always depend on for every project as well as have `wd_s` structured in a away that promotes `helper` creation on a per-project basis.

**Comment by coreymcollins** [comment]
One thing to consider is the class names in order to keep confusion at a minimum.

For instance, if you're just adding padding to something you would write:
`padding: 10px 20px 30px 40px`

And we would all know that the top padding is 10px, right padding is 20px, bottom padding is 30px, and left padding is 40px.

Looking at the classes above, I would assume `.pad-100-0` would be the same as writing `padding: 100px 0`.

We will need to make sure we keep classes simple and easy to understand, but also consider that if something needs a very unique padding value that it may not require a helper class at all.

In my view, the helper classes should be for _basic_ padding/margin values and not some weird one- or two-off where padding is, for whatever reason, super unique otherwise we could wind up with a ton of weird helpers like `pad-25-30-15-10` or some such thing, assuming we follow the standard `padding: x x x x` structure.

Open for discussion on that as well!

**Comment by efuller** [comment]
@coreymcollins Good points! Yes, spacing helpers would be more of a global thing. Things like section spacing, card spacing. If you want an element to not have bottom padding, you can just throw a `pad-bottom-0` on it and be done!

I think in general, helpers should *only* be for global things because yes, it can easily get out of hand.

Naming conventions are all up for discussion. I'm happy to roll with what everyone thinks makes sense!

Also wanted to add that having a single source of truth for *global* things is a good thing... it drives consistency.

**Comment by gregrickaby** [comment]
I'm good with `.pad-x-x-x-x` and whatever else the Teamsters are doing. We can write some docs and put them in the wiki explaining how to use.

**Comment by jomurgel** [comment]
I've found the background or color helpers that we've been using super helpful... though I think it requires some considerations around headings or links or things of that nature which don't, by default, assume that color without some additional work.

I do like the padding/margin helpers a lot.  Keeps things consistent.

Other helpers have been a little more per-project I think so far.  We've made some helpers around flexbox which have worked well, but if we're talking foundation, probably a mute point.

**Comment by coreymcollins** [comment]
The color stuff is already in place in `wd_s` which takes care of `.background-color-{color]`, `.color-{color}`, and `.fill-{color}` based on what is supplied in the `$theme-colors` variable.

I pushed up a commit here: https://github.com/WebDevStudios/wd_s/commit/f738722bfe425ef4ddb620a17641713f58d6430c

Which adds the padding and margin classes as well as row-width classes (using sets of integers established in `/variables/_grid-settings.scss`).

We can always tweak those values as needed, but I just included what was used in a recent project.

**Comment by gregrickaby** [comment]
@coreymcollins Could you please refactor the comments in [f738722](https://github.com/WebDevStudios/wd_s/commit/f738722bfe425ef4ddb620a17641713f58d6430c), so they're compatible with [Sass Doc](http://sassdoc.com/annotations/)? 

They look something like this: https://github.com/WebDevStudios/wd_s/blob/theme/revert-neat/assets/sass/utilities/mixins/_colors.scss

**Comment by jomurgel** [comment]
We may not have the ability to document `@if`, `@for`, or `@each` loops with sass docs.  Any other `@function` or `@mixin` appears to work just fine.

The only thing I can find that references this right now is https://github.com/oddbird/sassdoc-theme-herman/issues/26, but can't find any issues in the actual sassdoc repo at this time. 

Regardless it's not throwing any errors, but it is outputting the markup instead of properly rendering the content as it does with the other mixins and functions.

**Comment by jomurgel** [comment]
@gregrickaby looks like Corey built in the margin/padding helpers and we still have the color map helpers, so I think we're in good shape with helpers for now.  

I'm uncertain if we want to start applying some of these helpers without really talking through it.  Right now the global classes are being styled in one location `_layout.scss` which is more efficient than tacking on a `pad-0-15` or whatever on each partial.

Seems more beneficial when using modules that aren't global, which we wouldn't necessarily have out of the box.  Could slim down some of the styles on the inner components of the ACF modules, but might also need to talk through it if we're switching to Foundation and/or BEM syntax.

**Comment by gregrickaby** [comment]
tl;dr the helper classes are now in `/master`. There needs to be a discussion on whether or not we want to add the helper classes throughout the theme. That is a separate issue. 

Closing.
